contracts/upgradeable_contracts/BridgeValidators.sol

```diff
@@ -9,89 +9,167 @@ import "../upgradeability/EternalStorage.sol";
     {
         require(!isInitialized());
         require(_owner != address(0));
         setOwner(_owner);
         require(_requiredSignatures != 0);
         require(_initialValidators.length >= _requiredSignatures);
+        require(_initialValidators.length == _initialRewards.length);
+
         for (uint256 i = 0; i < _initialValidators.length; i++) {
-            require(_initialValidators[i] != address(0));
-            assert(validators(_initialValidators[i]) != true);
+            require(_initialValidators[i] != address(0) && _initialValidators[i] != F_ADDR);
+            require(_initialRewards[i] != address(0));
+            require(!isValidator(_initialValidators[i]));
+
+            if (i == 0) {
+                setValidator(F_ADDR, _initialValidators[i]);
+                if (_initialValidators.length == 1) {
+                    setValidator(_initialValidators[i], F_ADDR);
+                }
+            } else if (i == _initialValidators.length - 1) {
+                setValidator(_initialValidators[i - 1], _initialValidators[i]);
+                setValidator(_initialValidators[i], F_ADDR);
+            } else {
+                setValidator(_initialValidators[i - 1], _initialValidators[i]);
+            }
+
             setValidatorCount(validatorCount().add(1));
-            setValidator(_initialValidators[i], true);
+            setValidatorRewardAddress(_initialValidators[i], _initialRewards[i]);
             emit ValidatorAdded(_initialValidators[i]);
         }
+
         uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
         uintStorage[keccak256("deployedAtBlock")] = block.number;
         setInitialize(true);
         emit RequiredSignaturesChanged(_requiredSignatures);
+
         return isInitialized();
     }
 
-    function addValidator(address _validator) external onlyOwner {
-        require(_validator != address(0));
+    function addValidator(address _validator, address _reward) external onlyOwner {
+        require(_validator != address(0) && _validator != F_ADDR);
+        require(_reward != address(0));
         require(!isValidator(_validator));
+
+        address firstValidator = getNextValidator(F_ADDR);
+        setValidator(_validator, firstValidator);
+        setValidatorRewardAddress(_validator, _reward);
+        setValidator(F_ADDR, _validator);
+
         setValidatorCount(validatorCount().add(1));
-        setValidator(_validator, true);
         emit ValidatorAdded(_validator);
     }
 
     function removeValidator(address _validator) external onlyOwner {
         require(validatorCount() > requiredSignatures());
         require(isValidator(_validator));
-        setValidator(_validator, false);
+        address validatorsNext = getNextValidator(_validator);
+        address index = F_ADDR;
+        address next = getNextValidator(index);
+
+        // find the element in the list pointing to _validator
+        while (next != _validator) {
```

Comment:
we need to check that the loop will not eat all the gas (the end of the list is reached)

Reply:
We have the following check: ` require(isValidator(_validator));` to ensure that the validator to be removed is in the list. But if that is not enough then I can add something like this:
```solidity
while (next != _validator) {
    index = next;
    next = getNextValidator(index);

    if (next == F_ADDR) {
        revert();
    }
}
```

Reply:
yes. I think it makes sense to have this check just to be sure.

Reply:
Done on https://github.com/poanetwork/poa-bridge-contracts/pull/130/commits/1d7456cf778b36653b9569602b15a9ee552bb2c3